plutus-core/cost-model/budgeting-bench/Benchmarks/Values.hs

```diff
@@ -0,0 +1,115 @@
+    ]
+
+newtype PolicyId = PolicyId ByteString
+newtype TokenName = TokenName ByteString
+newtype Amount = Amount Integer
+type Counter = Integer
+
+data GenState = GenState
+    { policyIdCounter  :: Counter
+    , tokenNameCounter :: Counter
+    }
+
+type BenchState = StateT StdGen (State GenState)
+
+-- | An insertCoin benchmark is a concrete set of arguments we apply to the
+-- InsertCoin builtin function to measure its runtime cost.
+data InsertCoinBenchmark = InsertCoinBenchmark
+    { icPolicyId  :: PolicyId
+    , icTokenName :: TokenName
+    , icAmount    :: Amount
+    , icValue     :: Value
+    }
+
+icToRawTuple :: InsertCoinBenchmark -> (ByteString, ByteString, Integer, Value)
+icToRawTuple (InsertCoinBenchmark (PolicyId p) (TokenName t) (Amount a) v) = (p, t, a, v)
+
+benchInsertCoin :: StdGen -> Benchmark
+benchInsertCoin gen =
+    createFourTermBuiltinBenchElementwiseWithWrappers
+        (id, id, id, id) -- TODO: use proper wrappers
+        InsertCoin
+        []
+        (icToRawTuple <$> insertCoinBenchGen gen)
+
+-- | Generate a set of benchmarks for the InsertCoin builtin function.
+-- It includes the following scenarios:
+--   1. Inserting into an empty Value.
+--   2. Inserting a new TokenName into an existing PolicyId. Randomly extracting a PolicyId from the Value.
+--   3. Inserting into an existing TokenName. Randomly extracting a (PolicyId, TokenName) pair from the Value.
+--   4. Inserting a new PolicyId.
+--   5. Deleting a TokenName by inserting a 0 amount. Randomly extracting a (PolicyId, TokenName) pair from the Value.
+--   6. Deleting a PolicyId by inserting a 0 amount into its last TokenName. Should generate a Value with multiple such PolicyIds, and randomly picking which PolicyId to delete.
+-- We're interested in the worst case performance, so we'll use the largest key values possible.
+-- We should also run randomized benchmarks, where we insert random values into random Values.
+-- We actually want to see how the performance scales with the size of the Value, so we should generate Values of varying sizes.
+-- We want to make sure we are also hitting the worst case scenarios and various edge cases.
+insertCoinBenchGen
+    :: StdGen
+    -> [InsertCoinBenchmark]
+insertCoinBenchGen g = flip evalState (GenState 0 0) $ runStateGenT_ g $ \gen -> do
+    policyId <- newPolicyId gen
+    tokenName <- newTokenName gen
+    amount <- uniformAmount gen
+    let emptyValueBench = InsertCoinBenchmark policyId tokenName amount Value.empty
+    pure [emptyValueBench]
+
+-- | Generate a unique PolicyId on a uniform distribution. Note that the size of the
+-- generated bytestring is going to be larger than Value.maxKeyLen, because we
+-- append a counter integer to ensure uniqueness. This is acceptable for benchmarking
+-- purposes, as we're interested in the worst-case performance.
+newPolicyId :: StateGenM StdGen -> BenchState PolicyId
+newPolicyId gen = do
+    bs <- uniformByteStringM Value.maxKeyLen gen
```

Comment:
It's better to fix the first 30 bytes, and randomize the last 2 bytes. That aligns better with worst case. I think that's what Yura's generator does, so you can probably reuse that.